---
title: "Comparative analysis of gene expression between DMSO and PTX treatments"
author: "Anagha Shenoy"
date: "`r Sys.Date()`"
output:
  html_document:
      toc: true
      toc_depth: 3
      theme: cerulean
      highlights: kate
      df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This data was generated by the Heiser Lab at OHSU and is available at [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE266934). 

> HCC1143 cells were grown in vitro and treated for 24 or 72 hours with soluble perturbation.

# Load packages and data

```{r pkg}

library(Seurat)
library(SeuratObject)

library(dplyr)
library(pheatmap)
library(patchwork)
library(ggplot2)
library(cowplot)
library(future)
library(tidyverse)
library(presto) # devtools::install_github('immunogenomics/presto')

plan(multisession, workers = 4) # Set up parallelization

```

Data is located in `../data`, in folders corresponding to the treatment. 

`LH7200` is DMSO (control) treatment, and will hereafter be referred to as DMSO.

`LH7201` is paclitaxel (drug) treatment, and will hereafter be referred to as PTX.

*Note: This data has been pre-processed. If using raw data, quality control and correcting for batch effects is a necessary step. This helps refine the analysis, excluding cells that would affect the results: doublets (cells identified with too many features) and dead cells (cells identified to have high mitochondrial content).*

```{r}

LH7200.data <- Read10X(data.dir = "../data/LH7200/filtered_feature_bc_matrix")
LH7201.data <- Read10X(data.dir = "../data/LH7201/filtered_feature_bc_matrix")

LH7200 <- CreateSeuratObject(LH7200.data, project="LH7200") # control (DMSO)
LH7201 <- CreateSeuratObject(LH7201.data, project="LH7201") # treatment (Paclitaxel)

```

# Generate exploratory plots

Here we visualize the number of features (genes) and count of RNA (number of molecules) per cell, in each of the DMSO and PTX treatment groups.

## DMSO

```{r}

LH7200_meta <- LH7200@meta.data

num_features_LH7200 <- ggplot(LH7200_meta, aes(x = nFeature_RNA)) +
                        geom_histogram(bins=75, color="cornflowerblue") +
                        theme_minimal() +
                        labs(title = "Number of Genes per Cell, DMSO", 
                             x = "Number of Genes", 
                             y = "Frequency")

num_count_LH7200 <- ggplot(LH7200_meta, aes(x = nCount_RNA)) +
                        geom_histogram(bins=75, color="cornflowerblue") +
                        theme_minimal() +
                        labs(title = "Number of Molecules per Cell, DMSO", 
                             x = "Number of Molecules", 
                             y = "Frequency")

num_features_LH7200 + num_count_LH7200

```

## PTX

```{r}

LH7201_meta <- LH7201@meta.data

num_features_LH7201 <- ggplot(LH7201_meta, aes(x = nFeature_RNA)) +
                        geom_histogram(bins=75, color="cornflowerblue") +
                        theme_minimal() +
                        labs(title = "Number of Genes per Cell, PTX", 
                             x = "Number of Genes", 
                             y = "Frequency")

num_count_LH7201 <- ggplot(LH7201_meta, aes(x = nCount_RNA)) +
                        geom_histogram(bins=75, color="cornflowerblue") +
                        theme_minimal() +
                        labs(title = "Number of Molecules per Cell, PTX", 
                             x = "Number of Molecules", 
                             y = "Frequency")

num_features_LH7201 + num_count_LH7201

```

## Merged data

To conduct differential gene expression analysis, we merge the datasets, labeling each of the groups appropriately with "CTRL" (DMSO) and "TRT" (PTX).

```{r}

LH72 <- merge(LH7200, 
              LH7201, 
              add.cell.ids = c("CTRL", "TRT"), 
              project = "LH")

head(colnames(LH72))

LH72$stim <- LH72$orig.ident

table(LH72$stim)

```

```{r}

LH72[["percent_mt"]] <- PercentageFeatureSet(LH72, pattern = "^MT-")

VlnPlot(LH72, features = c("nFeature_RNA", "nCount_RNA", "percent_mt"), ncol = 3)

```
```{r}

percent_mt_visual <- FeatureScatter(LH72, 
                                    feature1 = "nCount_RNA", 
                                    feature2 = "percent_mt") + 
                      theme(text = element_text(size = 20))

nfeature_visual <- FeatureScatter(LH72, 
                                  feature1 = "nCount_RNA", 
                                  feature2 = "nFeature_RNA") + 
                      theme(text = element_text(size = 20))

percent_mt_visual + nfeature_visual

```

# Conduct differential gene expression analysis

We first need to normalize the count data (for each layer / treatment group).

```{r}

LH72 <- NormalizeData(LH72)

```

Find and plot highly variable genes.

```{r}

LH72 <- FindVariableFeatures(LH72)

top10_hv_genes <- head(VariableFeatures(LH72), 10) # highly variable genes

features <- VariableFeaturePlot(LH72)

features_labeled <- LabelPoints(plot = features, 
                                points = top10_hv_genes, 
                                repel = TRUE) + 
                    NoLegend()

features_labeled + ggtitle("Cell-to-cell variation within LH72")

```

We scale and perform PCA on the data in order to reduce dimensionality.

```{r}

LH72 <- ScaleData(LH72)

LH72 <- RunPCA(LH72)

```

Then, cluster the data.

```{r}

LH72 <- FindNeighbors(LH72, 
                      dims = 1:30,
                      reduction = "pca")

LH72 <- FindClusters(LH72, 
                     resolution = 2, 
                     cluster.name = "unintegrated_clusters")

```

With UMAP, we conduct non-linear dimensional reduction.

```{r}

LH72 <- RunUMAP(LH72, 
                dims = 1:30, 
                reduction = "pca", 
                reduction.name = "umap.unintegrated")

DimPlot(LH72, 
        reduction = "umap.unintegrated", 
        group.by = c("stim", "seurat_clusters"))

```

## Integrate data

We need to integrate/align data from the two conditions (control and treatment).

```{r}

LH72 <- IntegrateLayers(object = LH72, 
                        method = CCAIntegration, 
                        orig.reduction = "pca", 
                        new.reduction = "integrated.cca",
                        verbose = FALSE)

LH72[["RNA"]] <- JoinLayers(LH72[["RNA"]])

LH72 <- FindNeighbors(LH72, 
                      reduction = "integrated.cca", 
                      dims = 1:30)

LH72 <- FindClusters(LH72, 
                     resolution = 1)

```

```{r}

LH72 <- RunUMAP(LH72, 
                dims = 1:30, 
                reduction = "integrated.cca")

DimPlot(LH72, 
        reduction = "umap", 
        split.by = "stim")

```

Find markers conserved between the two treatment groups.

```{r}

Idents(LH72) <- "seurat_clusters"

conserved_markers <- FindConservedMarkers(LH72, 
                                          ident.1 = 0, 
                                          grouping.var = "stim", 
                                          verbose = FALSE)

head(conserved_markers)

```

Plot select conserved genes across conditions.

```{r}

select_conserved <- rownames(conserved_markers)[1:3]

FeaturePlot(LH72, 
            reduction = "umap", 
            features = select_conserved,
            split.by = "stim")

```

Plot select highly variable genes across conditions.

```{r}

select_hv <- top10_hv_genes[1:3]

FeaturePlot(LH72, 
            reduction = "umap", 
            features = select_hv,
            split.by = "stim")

```

## Identify differentially expressed genes

Use the `FindMarkers` function to identify differentially expressed genes for each identity class.

```{r}

LH72$celltype.stim <- paste(LH72$seurat_clusters, 
                            LH72$stim, 
                            sep = "_")

Idents(LH72) <- "celltype.stim"

diff_expressed <- FindMarkers(LH72, 
                              ident.1 = "0_LH7200",
                              ident.2 = "0_LH7201", 
                              verbose = FALSE)

head(diff_expressed, n = 15)

```

Plot select differentially expressed genes identified in the previous step.

```{r}

FeaturePlot(LH72, 
            features = c("IFI6", "B2M", "HLA-B"), 
            split.by = "stim", 
            max.cutoff = 3, 
            cols = c("grey", "red"), 
            reduction = "umap")

```

Visualize expression levels for genes identified to be differentially expressed.

```{r}

violin_hv <- VlnPlot(LH72, 
                     features = c("IFI6", "B2M", "HLA-B"), 
                     split.by = "stim", 
                     group.by = "seurat_clusters",
                     pt.size = 0, 
                     combine = FALSE)

violin_hv[[1]] <- violin_hv[[1]] + theme(axis.title.y=element_text(size = 8),
                                         axis.text.y=element_text(size = 8))
violin_hv[[2]] <- violin_hv[[2]] + theme(axis.title.y=element_text(size = 8),
                                         axis.text.y=element_text(size = 8))
violin_hv[[3]] <- violin_hv[[3]] + theme(axis.title.y=element_text(size = 8),
                                         axis.text.y=element_text(size = 8))
             

wrap_plots(plots = violin_hv, ncol = 1)
 

```

Use the `FindAllMarkers` function to identify differentially expressed genes across the two conditions.

### Generate heatmap

```{r}

Idents(LH72) <- "stim"

cluster.markers <- FindMarkers(LH72, 
                               ident.1 = "LH7200", 
                               ident.2 = "LH7201")

cluster.markers

```

```{r}

LH72.markers <- FindAllMarkers(LH72, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

top10_markers <- LH72.markers %>%
                  group_by(cluster) %>%
                  top_n(n = 10, wt = avg_log2FC)

DoHeatmap(LH72, angle=30, group.colors=c("gray", "black"), draw.lines=TRUE, features = top10_markers$gene)

```

### Generate correlation matrix

```{r}

genes.to.correlate <- top10_markers$gene

correlation.matrix <- FetchData(LH72, 
                                vars = genes.to.correlate) %>% 
                      cor()

pheatmap(correlation.matrix)

```

### Generate dot plot

```{r}

DotPlot(LH72, 
        features=unique(top10_markers$gene), 
        dot.scale = 4) + 
  
      coord_flip() + 
  
      labs(title = "Differential Gene Expression\nDMSO vs. PTX at 72 Hours", 
           x = "Genes", 
           y = "Treatment Group") + 
  
      theme(legend.text = element_text(size = 10), 
            legend.title = element_text(size = 10),
            axis.text.y = element_text(size = 8))

```

